samza-core/src/main/java/org/apache/samza/execution/ExecutionPlanner.java

```diff
@@ -188,98 +183,82 @@ private void validateConfig() {
-            throw  new SamzaException(String.format(
-                "Unable to resolve input partitions of stream %s for the join %s. Expected: %d, Actual: %d",
-                edge.getName(), join.getOpId(), partitions, edgePartitions));
-          }
-        }
-      }
-
-      // assign the partition count for intermediate streams
-      for (StreamEdge edge : joinSpecToStreamEdges.get(join)) {
-        if (edge.getPartitionCount() <= 0) {
-          log.info("Set the partition count to {} for input stream {} to the join {}.",
-              new Object[] {partitions, edge.getName(), join.getOpId()});
-          edge.setPartitionCount(partitions);
-
-          // find other joins can be inferred by setting this edge
-          for (JoinOperatorSpec op : streamEdgeToJoinSpecs.get(edge)) {
-            if (!visited.contains(op)) {
-              joinQ.add(op);
-              visited.add(op);
-            }
-          }
-        }
-      }
+  private void validateJoinInputStreamPartitions(JobGraph jobGraph) {
+    // Group joined input operator specs (input/intermediate streams) by the join
+    // operator specs corresponding to the join operations they participate in.
+    Multimap<JoinOperatorSpec, InputOperatorSpec> joinOpSpecToInputOpSpecs =
+        OperatorSpecGraphAnalyzer.getJoinedInputOperatorSpecs(jobGraph.getSpecGraph());
+
+    // Convert every group of input operator specs into a group of corresponding stream edges.
+    List<JoinedStreamsGroup> joinedStreamsGroups = new ArrayList<>();
+    for (JoinOperatorSpec joinOpSpec : joinOpSpecToInputOpSpecs.keySet()) {
+      Collection<InputOperatorSpec> joinedInputOpSpecs = joinOpSpecToInputOpSpecs.get(joinOpSpec);
+      JoinedStreamsGroup joinedStreamsGroup = getJoinedStreamsGroup(joinOpSpec.getOpId(), joinedInputOpSpecs, jobGraph);
+      joinedStreamsGroups.add(joinedStreamsGroup);
     }
+
+    /*
+     * Sort the stream edge groups by their category so they appear in this order:
+     *   1. groups composed exclusively of stream edges with set partition counts
+     *   2. groups composed of a mix of stream edges  with set/unset partition counts
+     *   3. groups composed exclusively of stream edges with unset partition counts
+     *
+     *   This guarantees that we process the most constrained stream edge groups first,
+     *   which is crucial for intermediate stream edges that are members of multiple
+     *   stream edge groups. For instance, if we have the following groups of stream
+     *   edges (partition counts in parentheses, question marks for intermediate streams):
+     *
+     *      a. e1 (16), e2 (16)
+     *      b. e2 (16), e3 (?)
+     *      c. e3 (?), e4 (?)
+     *
+     *   processing them in the above order (most constrained first) is guaranteed to
+     *   yield correct assignment of partition counts of e3 and e4 in a single scan.
+     */
+    Collections.sort(joinedStreamsGroups, Comparator.comparing(JoinedStreamsGroup::getCategory));
```

Comment:
I came to the same conclusion actually — making `StreamEdge` immutable can greatly improve the `ExecutionPlanner`. I didn't want to do it in this series of PRs to avoid a scoop creep though.

I think one idea we can explore is making partition count a ctor param and a readonly property of a `StreamEdge`. This would require deferring the creation of any `StreamEdge` until its partition count is known, which is probably not going to be difficult after this PR.

I'll take note of this and send a follow-up PR later on.